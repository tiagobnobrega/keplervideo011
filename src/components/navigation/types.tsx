/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
