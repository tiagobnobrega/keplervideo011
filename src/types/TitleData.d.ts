import { MediaItem } from '@amzn/keplerblocks-types';

export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface TitleData extends MediaItem {
  videoUrl: string;
  categories: string[];
  channelID: string;
  posterUrl: string;
  format: string;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
  rentAmount: string;
}
