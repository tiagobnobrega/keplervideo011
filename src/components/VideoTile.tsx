import { isEqual } from 'lodash';
import React, { memo, RefObject, useRef } from 'react';
import { Animated, StyleSheet, Systrace, View } from 'react-native';

import {
  ContentIdNamespaces,
  ContentInteractionType,
  ContentPersonalizationServer,
} from '@amzn/kepler-content-personalization';
import { Card } from '@amzn/kepler-ui-components';
import { useNavigation } from '@amzn/react-navigation__native';
import { isContentPersonalizationEnabled } from '../config/AppConfig';
import {
  getMockContentID,
  getMockContentInteraction,
} from '../mocks/ContentPersonalizationMocks';
import { TitleData } from '../types/TitleData';
import { HomeScreenNavigationProps, Screens } from './navigation/types';

const FOCUSED_TILE_SCALE = 1.05;
const DEFAULT_TILE_SCALE = 1.0;
export interface VideoTileProps {
  index: number;
  data: TitleData;
  onFocus?: (title?: TitleData) => void;
  onBlur?: (title?: TitleData) => void;
  row?: number;
  firstElementRef?: RefObject<View> | null;
}

const VideoTile = ({
  data,
  onFocus,
  onBlur,
  index,
  row,
  firstElementRef,
}: VideoTileProps) => {
  const navigation = useNavigation<HomeScreenNavigationProps>();
  const focusableElementRef = useRef<View>(null);
  const animatedScale = useRef(new Animated.Value(DEFAULT_TILE_SCALE)).current;

  const reportContentNavigation = () => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }

    const contentInteraction = getMockContentInteraction(
      ContentInteractionType.DETAIL_VIEW,
      getMockContentID(data.title, ContentIdNamespaces.NAMESPACE_CDF_ID),
    );
    console.info(
      `k_content_per: Reporting new content interaction. Ingressing into Detail view for title : ${data.title}`,
    );
    ContentPersonalizationServer.reportNewContentInteraction(
      contentInteraction,
    );
  };

  const navigateToDetailsScreen = async () => {
    Systrace.beginEvent('nav:detail-screen');
    reportContentNavigation();

    navigation.push(Screens.DETAILS_SCREEN, {
      data: data,
      sendDataOnBack: handleFocusOnBack,
    });

    Systrace.endEvent();
  };

  const handleFocusOnBack = (id: number | string) => {
    if (id === data.id) {
      //if user clicked first tile then focus on first tile when user comes back
      if (index === 0 && row === 0) {
        firstElementRef?.current?.focus();
      } else {
        focusableElementRef.current?.focus();
      }
    }
  };

  const startScaledAnimation = (scale: number) => {
    Animated.timing(animatedScale, {
      toValue: scale,
      duration: 250,
      useNativeDriver: true,
    }).start();
  };

  const onFocusHandler = () => {
    startScaledAnimation(FOCUSED_TILE_SCALE);
    onFocus?.(data);
  };

  const firstTileBlur = () => {
    firstElementRef?.current?.blur();
  };

  const onBlurHandler = () => {
    startScaledAnimation(DEFAULT_TILE_SCALE);
    if (index === 0 && row === 0) {
      firstTileBlur();
    } else {
      focusableElementRef.current?.blur();
    }
    onBlur?.(data);
  };

  return (
    <Animated.View
      ref={index === 0 && row === 0 ? firstElementRef : focusableElementRef}
      testID={`tileContainer${data.id}`}
      style={[styles.container, { transform: [{ scale: animatedScale }] }]}
      // @ts-ignore // TS 2339: Property 'onFocus' does not exist on type 'Animated.View'.
      onFocus={onFocusHandler}
      onBlur={onBlurHandler}>
      <Card
        orientation={'vertical'}
        size="md"
        testID={`card${data.id}`}
        onPress={navigateToDetailsScreen}>
        <Card.Image
          source={{
            uri: data.thumbnail,
          }}
          edgeToEdge
        />
      </Card>
    </Animated.View>
  );
};

const dataUnchanged = (
  prevProps: VideoTileProps,
  nextProps: VideoTileProps,
) => {
  return isEqual(prevProps.data, nextProps.data);
};
export default memo(VideoTile, dataUnchanged);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
