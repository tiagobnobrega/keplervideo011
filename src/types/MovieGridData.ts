import { TitleData } from './TitleData';

export type MovieGridData = {
  heading: string;
  testID: string;
  data: () => TitleData[];
};
