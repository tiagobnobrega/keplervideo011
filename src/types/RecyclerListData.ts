import { StyleProp, ViewStyle } from 'react-native';
import { RecyclerListViewProps } from 'recyclerlistview';

export interface CustomRecyclerListViewProps extends RecyclerListViewProps {
  contentContainerStyle?: StyleProp<ViewStyle>;
  initialNumToRender: number | undefined;
}
