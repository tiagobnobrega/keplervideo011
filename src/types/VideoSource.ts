// adapted from GStreamer MIME Types (https://gstreamer.freedesktop.org/documentation/subparse/subparse.html?gi-language=c)
export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface VideoSource {
  title: string;
  uri: string;
  format: VideoFormat;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
}
