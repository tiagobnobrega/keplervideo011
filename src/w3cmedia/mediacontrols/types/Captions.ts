import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export interface CaptionButtonProps {
  onPress: () => void;
  video: VideoPlayer;
  captionMenuVisibility: boolean;
  testID: string;
  playerControlType?: string;
  isCaptionButtonFocused: boolean | null;
}

export interface CaptionOptionProps {
  video: VideoPlayer;
  enableCaption: (captionId: string) => void;
  selectedCaptionId: string | null;
}

export interface CaptionMenuItemProps {
  text: string;
  index?: string;
  selected?: boolean;
  onPress: () => void;
  testID?: string;
}

export interface CaptionMenuProps {
  captionMenuVisibility: boolean;
  video: VideoPlayer;
  setSelectedCaption: (id: string) => void;
}
