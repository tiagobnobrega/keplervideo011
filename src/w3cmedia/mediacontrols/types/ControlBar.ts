import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export enum PlayerControlType {
  PLAY = 'play',
  PAUSE = 'pause',
  PLAYPAUSE = 'playpause',
  SKIPBACKWARD = 'skip_backward',
  SKIPFORWARD = 'skip_forward',
}
export interface ControlBarButtonProps {
  videoRef: React.MutableRefObject<VideoPlayer | null>;
  onBlur?: () => void;
  playerControlType?: string;
}
