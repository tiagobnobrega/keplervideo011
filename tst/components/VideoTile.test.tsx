/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { jest as jestGlobals } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import * as ReactNative from 'react-native';

import { render } from '@testing-library/react-native';
import React from 'react';
import VideoTile from '../../src/components/VideoTile';
import { TitleData } from '../../src/types/TitleData';

jestGlobals.mock('../../src/components/navigation/AppStack', () => ({
  __esModule: true,

  Screen: { DETAILS_SCREEN: 'detail-screen' },
}));

jestGlobals.mock('../../src/mocks/ContentPersonalizationMocks', () => ({
  __esModule: true,

  getMockContentID: jest.fn(),
  getMockContentInteraction: jest.fn(),
}));

jestGlobals.mock('@amzn/react-navigation__native', () => ({
  __esModule: true,

  useNavigation: jest.fn(),
}));

jestGlobals.mock('@amzn/kepler-content-personalization', () => ({
  __esModule: true,

  ContentPersonalizationServer: jest.fn(),
  ContentInteractionType: jest.fn(),
  ContentIdNamespaces: jest.fn(),
}));

const createVideoTitle = (
  id: string,
  title: string,
  description: string,
): TitleData => {
  return {
    id: id,
    title: title,
    description: description,
    duration: 86,
    thumbnail: 'https://example.com/thumbnail.jpg',
    posterUrl: '',
    videoUrl: 'https://example.com/video.mp4',
    categories: ['Category 1', 'Category 2'],
    releaseDate: '2023-01-01',
    channelID: 'XXXXX',
    rating: '4.5',
    mediaType: 'video',
    mediaSourceType: 'url',
    format: 'MP4',
    secure: false,
    uhd: true,
    rentAmount: '',
  };
};

describe('VideoTile tests', () => {
  jestGlobals.useFakeTimers();

  // @ts-ignore ts(2540): Animated.timing is read-only, but we
  // must overwrite it to make it work for our test.
  ReactNative.Animated.timing = (value: any, config: any) => {
    return {
      start: (callback: any) => {
        value.setValue(config.toValue);
        callback && callback();
      },
    };
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot on first render', async () => {
    const title = createVideoTitle('1', 'Title 1', 'Description 1');
    const renderedTile = render(<VideoTile index={0} data={title} />);

    expect(renderedTile.toJSON()).toMatchSnapshot();
  });
});
