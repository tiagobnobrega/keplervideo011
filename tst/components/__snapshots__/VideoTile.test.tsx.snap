// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoTile tests should match the snapshot on first render 1`] = `
<View
  collapsable={false}
  onBlur={[Function]}
  onFocus={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
  testID="tileContainer1"
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#202225",
          "borderColor": "transparent",
          "borderRadius": 16,
          "borderWidth": 5,
          "height": 314,
          "margin": undefined,
          "overflow": "hidden",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 2,
          "width": 214,
        },
        Object {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "overflow": "hidden",
        },
        undefined,
        false,
      ]
    }
    testID="card1"
  >
    <View
      style={
        Object {
          "aspectRatio": 1,
          "height": undefined,
          "width": 204,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://example.com/thumbnail.jpg",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 0,
            },
            Object {
              "height": "100%",
              "resizeMode": "cover",
              "width": "100%",
            },
            undefined,
            false,
          ]
        }
      />
    </View>
  </View>
</View>
`;
